heft-plugins/heft-serverless-stack-plugin/src/ServerlessStackPlugin.ts

```diff
@@ -0,0 +1,229 @@
+            return null;
+          }
+        );
+
+        bundle.hooks.run.tapPromise(PLUGIN_NAME, async () => {
+          if (!sstParameter.value) {
+            this._logger.terminal.writeLine(
+              'Skipping SST operations because the "--sst" command-line parameter was not specified.'
+            );
+          } else {
+            await this._onBundleRunAsync({
+              heftSession,
+              heftConfiguration,
+              serveMode: build.properties.serveMode,
+              sstCliEntryPoint,
+              sstCliPackagePath,
+              sstStageParameter
+            });
+          }
+        });
+      });
+    });
+  }
+
+  private async _onBundleRunAsync(options: {
+    heftSession: HeftSession;
+    heftConfiguration: HeftConfiguration;
+    serveMode: boolean;
+    sstCliEntryPoint: string;
+    sstCliPackagePath: string;
+    sstStageParameter: IHeftStringParameter;
+  }): Promise<void> {
+    const sstCommandArgs: string[] = [];
+    sstCommandArgs.push(options.sstCliEntryPoint);
+
+    if (options.serveMode) {
+      sstCommandArgs.push('start');
+    } else {
+      sstCommandArgs.push('build');
+    }
+    if (options.heftSession.debugMode) {
+      sstCommandArgs.push('--verbose');
+    }
+    if (options.sstStageParameter.value) {
+      sstCommandArgs.push('--stage');
+      sstCommandArgs.push(options.sstStageParameter.value);
+    }
+
+    this._logger.terminal.writeVerboseLine('Launching child process: ' + JSON.stringify(sstCommandArgs));
+
+    const sstCommandEnv: NodeJS.ProcessEnv = this._getWorkaroundEnvironment(options.sstCliPackagePath);
+
+    const sstCommandResult: child_process.ChildProcess = child_process.spawn(
+      process.execPath,
+      sstCommandArgs,
+      {
+        cwd: options.heftConfiguration.buildFolder,
+        stdio: ['inherit', 'pipe', 'pipe'],
+        env: sstCommandEnv,
+        ...SubprocessTerminator.RECOMMENDED_OPTIONS
+      }
+    );
+
+    SubprocessTerminator.killProcessTreeOnExit(sstCommandResult, SubprocessTerminator.RECOMMENDED_OPTIONS);
+
+    let completionResolve: () => void;
+    let completionReject: (reason: Error) => void;
+
+    const completionPromise: Promise<void> = new Promise((resolve, reject) => {
+      completionResolve = resolve;
+      completionReject = reject;
+    });
+
+    sstCommandResult.stdout?.on('data', (chunk: Buffer) => {
+      this._writeOutput(chunk.toString(), (x) => this._logger.terminal.write(x));
+    });
+    sstCommandResult.stderr?.on('data', (chunk: Buffer) => {
+      this._writeOutput(chunk.toString(), (x) => this._logger.terminal.writeError(x));
+    });
+
+    sstCommandResult.on('exit', (code: number | null) => {
+      if (options.serveMode) {
+        // The child process is not supposed to terminate in watch mode
+        this._logger.terminal.writeErrorLine(`SST process terminated with exit code ${code}`);
+        // TODO: Provide a Heft facility for this
+        process.exit(1);
```

Comment:
If you throw, it should terminate the process gracefully.

Reply:
@iclanton this is the Node.js `ChildProcess.on()` event handler, not a Heft callstack